include/bl_storage.h

```diff
@@ -7,15 +7,63 @@
 #ifndef BL_STORAGE_H_
 #define BL_STORAGE_H_
 
-#include <zephyr/types.h>
 #include <string.h>
+#include <zephyr/types.h>
+#include <drivers/nrfx_common.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
 
+#ifdef CONFIG_IS_SECURE_BOOTLOADER
+#define PROVISION_PARTITION_ADDRESS PM_PROVISION_ADDRESS
+#else
+#define PROVISION_PARTITION_ADDRESS CONFIG_SECURE_BOOT_STORAGE_START_ADDRESS
+#endif
+
 #define EHASHFF 113 /* A hash contains too many 0xFs. */
+#define EREADLCS 114 /* LCS field of OTP is in an invalid state */
+#define EINVALIDLCS 115 /* Invalid LCS*/
+
+/** Storage for the PRoT Security Lifecycle state, that consists of 4 states:
+ *  - Device assembly and test
+ *  - PRoT Provisioning
+ *  - Secured
+ *  - Decommissioned
+ *  These states are transitioned top down during the life time of a device.
+ *  Therefore when setting a new LCS we just mark the old state as left, this
+ *  way one field less can be used to encode the lcs.
+ *  This works as ASSEMBLY implies the OTP be erased except of needed
+ *  key material.
+ */
+struct life_cycle_state_data {
+	uint16_t left_assembly;
+	uint16_t left_provisioning;
+	uint16_t left_secure;
+	/* Pad to end the alignment at a 4-byte boundary as the UICR->OTP
+	 * only supports 4-byte reads */
+	uint16_t reserved_for_padding;
+};
+
+/** The first data structure in the bootloader storage. It has unknown length
+ *  since 'key_data' is repeated. This data structure is immediately followed by
+ *  struct counter_collection.
+ */
+struct bl_storage_data {
+	/* NB: When placed in OTP, reads must be 4 bytes and 4 byte aligned */
+	struct life_cycle_state_data lcs;
+	uint32_t s0_address;
+	uint32_t s1_address;
+	uint32_t num_public_keys; /* Number of entries in 'key_data' list. */
+#ifdef CONFIG_SB_IMPLEMENTATION_ID
+	uint8_t implementation_id[64];
```

Comment:
Implementation ID is 32 bytes.
```suggestion
	uint8_t implementation_id[32];
```